File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/QueueEntryDataBinder.java
package com.github.bjoernpetersen.q.ui.fragments;

import android.support.annotation.Nullable;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import com.github.bjoernpetersen.jmusicbot.client.ApiException;
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry;
import com.github.bjoernpetersen.jmusicbot.client.model.Song;
import com.github.bjoernpetersen.q.QueueState;
import com.github.bjoernpetersen.q.R;
import com.github.bjoernpetersen.q.api.ApiKey;
import com.github.bjoernpetersen.q.api.Auth;
import com.github.bjoernpetersen.q.api.AuthException;
import com.github.bjoernpetersen.q.api.Connection;
import com.github.bjoernpetersen.q.api.Permission;
import com.squareup.picasso.Callback.EmptyCallback;
import com.squareup.picasso.Picasso;
import com.yqritc.recyclerviewmultipleviewtypesadapter.DataBindAdapter;
import com.yqritc.recyclerviewmultipleviewtypesadapter.DataBinder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Objects;


public class QueueEntryDataBinder extends DataBinder<QueueEntryDataBinder.ViewHolder> {

  private final QueueEntryListener listener;
  private final List<QueueEntry> items;

  public QueueEntryDataBinder(DataBindAdapter adapter, QueueEntryListener listener) {
    super(adapter);
    this.listener = listener;
    this.items = new ArrayList<>();
  }

  @Override
  public ViewHolder newViewHolder(ViewGroup parent) {
    View view = LayoutInflater.from(parent.getContext())
        .inflate(R.layout.fragment_queue, parent, false);
    return new ViewHolder(view);
  }

  @Override
  public void bindViewHolder(final ViewHolder holder, int position) {
    final QueueEntry entry = getItem(position);
    if (Objects.equals(entry, holder.entry)) {
      return;
    }
    holder.entry = entry;
    Song song = entry.getSong();

    final ImageView albumArtView = holder.albumArtView;
    Picasso.with(albumArtView.getContext()).cancelRequest(holder.albumArtView);
    holder.albumArtView.setVisibility(View.GONE);
    Picasso.with(albumArtView.getContext())
        .load(song.getAlbumArtUrl())
        .into(holder.albumArtView, new EmptyCallback() {
          @Override
          public void onSuccess() {
            albumArtView.setVisibility(View.VISIBLE);
          }
        });
    setContent(holder.titleView, song.getTitle());
    setContent(holder.descriptionView, song.getDescription());
    setContent(holder.durationView, getDurationString(song.getDuration()));
    setContent(holder.queuerView, entry.getUserName());

    holder.view.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        if (null != listener) {
          // Notify the active callbacks interface (the activity, if the
          // fragment is attached to one) that an item has been selected.
          listener.onClick(holder.entry);
        }
      }
    });

    final PopupMenu menu = new PopupMenu(holder.view.getContext(), holder.contextMenu);
    menu.inflate(R.menu.query_item_menu);
    menu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
      @Override
      public boolean onMenuItemClick(MenuItem item) {
        switch (item.getItemId()) {
          case R.id.remove_button:
            new Thread(new Runnable() {
              @Override
              public void run() {
                try {
                  String token = Auth.INSTANCE.getApiKey().getRaw();
                  Song song = entry.getSong();
                  List<QueueEntry> newQueue = Connection.INSTANCE.dequeue(
                      token, song.getId(), song.getProvider().getId()
                  );
                  QueueState.getInstance().set(newQueue);
                } catch (ApiException e) {
                  if (e.getCode() == 403) {
                    Auth.INSTANCE.clear();
                  } else {
                    e.printStackTrace();
                    // TODO show toast
                  }
                } catch (AuthException e) {
                  e.printStackTrace();
                  // TODO show toast
                }
              }
            }).start();
            return true;
          default:
            return false;
        }
      }
    });
    holder.contextMenu.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        MenuItem removeButton = menu.getMenu().findItem(R.id.remove_button);
        Auth auth = Auth.INSTANCE;
        ApiKey apiKey = auth.getApiKeyNoRefresh();
        removeButton.setVisible(apiKey != null && apiKey.getUserName().equals(entry.getUserName())
            || auth.hasPermissionNoRefresh(Permission.SKIP));
        menu.show();
      }
    });
  }

  @Nullable
  private String getDurationString(int durationSeconds) {
    if (durationSeconds == 0) {
      return null;
    }
    int seconds = durationSeconds % 60;
    int minutes = (durationSeconds - seconds) / 60;
    return String.format(Locale.US, "%d:%02d", minutes, seconds);
  }

  private void setContent(TextView view, String content) {
    if (content == null || content.isEmpty()) {
      view.setVisibility(View.GONE);
    } else {
      view.setVisibility(View.VISIBLE);
      view.setText(content);
    }
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  QueueEntry getItem(int position) {
    return items.get(position);
  }

  public List<QueueEntry> getItems() {
    return Collections.unmodifiableList(items);
  }

  public void setItems(List<QueueEntry> items) {
    this.items.clear();
    this.items.addAll(items);
    notifyDataSetChanged();
  }

  static class ViewHolder extends RecyclerView.ViewHolder {

    public final View view;
    public final ImageView albumArtView;
    public final TextView titleView;
    public final TextView descriptionView;
    public final TextView durationView;
    public final TextView queuerView;
    public final ImageButton contextMenu;
    public QueueEntry entry;

    public ViewHolder(View view) {
      super(view);
      this.view = view;
      this.albumArtView = (ImageView) view.findViewById(R.id.album_art);
      this.titleView = (TextView) view.findViewById(R.id.song_title);
      this.titleView.setSelected(true);
      this.descriptionView = (TextView) view.findViewById(R.id.song_description);
      this.descriptionView.setSelected(true);
      this.durationView = (TextView) view.findViewById(R.id.song_duration);
      this.queuerView = (TextView) view.findViewById(R.id.song_queuer);
      this.contextMenu = (ImageButton) view.findViewById(R.id.context_menu);
    }

    @Override
    public String toString() {
      return super.toString() + " '" + titleView.getText() + "'";
    }
  }

  public interface QueueEntryListener {

    void onClick(QueueEntry entry);
  }
}

File: app/src/main/java/com/github/bjoernpetersen/q/ui/fragments/QueueEntryDataBinder.kt
package com.github.bjoernpetersen.q.ui.fragments

import android.content.Context
import android.content.Intent
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import com.github.bjoernpetersen.jmusicbot.client.ApiException
import com.github.bjoernpetersen.jmusicbot.client.model.QueueEntry
import com.github.bjoernpetersen.q.QueueState
import com.github.bjoernpetersen.q.R
import com.github.bjoernpetersen.q.api.Auth
import com.github.bjoernpetersen.q.api.AuthException
import com.github.bjoernpetersen.q.api.Connection
import com.github.bjoernpetersen.q.api.Permission
import com.github.bjoernpetersen.q.api.action.MoveSong
import com.github.bjoernpetersen.q.ui.SearchActivity
import com.github.bjoernpetersen.q.ui.asDuration
import com.squareup.picasso.Callback.EmptyCallback
import com.squareup.picasso.Picasso
import com.yqritc.recyclerviewmultipleviewtypesadapter.DataBindAdapter
import com.yqritc.recyclerviewmultipleviewtypesadapter.DataBinder
import kotlinx.android.synthetic.main.fragment_queue.view.*
import java.util.*

class QueueEntryDataBinder(adapter: DataBindAdapter, private val listener: QueueEntryListener?) :
    DataBinder<QueueEntryDataBinder.ViewHolder>(adapter) {

  private var _items: MutableList<QueueEntry> = ArrayList()
  var items: List<QueueEntry>
    get() = Collections.unmodifiableList(_items)
    set(value) {
      val items = _items
      items.clear()
      items.addAll(value)
      notifyDataSetChanged()
    }

  override fun newViewHolder(parent: ViewGroup): ViewHolder {
    val view = LayoutInflater.from(parent.context)
        .inflate(R.layout.fragment_queue, parent, false)
    return ViewHolder(view)
  }

  override fun bindViewHolder(holder: ViewHolder, position: Int) {
    val entry = getItem(position)
    if (entry == holder.entry) {
      return
    }
    holder.entry = entry
    val song = entry.song

    val albumArtView = holder.albumArtView
    Picasso.with(holder.context).cancelRequest(holder.albumArtView)
    holder.albumArtView.visibility = View.GONE
    Picasso.with(albumArtView.context)
        .load(song.albumArtUrl)
        .into(holder.albumArtView, object : EmptyCallback() {
          override fun onSuccess() {
            albumArtView.visibility = View.VISIBLE
          }
        })
    holder.titleView.setContent(song.title)
    holder.descriptionView.setContent(song.description)
    holder.durationView.setContent(song.duration.asDuration())
    holder.queuerView.setContent(entry.userName)

    holder.view.setOnClickListener {
      holder.entry?.let { listener?.onClick(it) }
    }

    val menu = PopupMenu(holder.view.context, holder.contextMenu)
    menu.inflate(R.menu.query_item_menu)
    menu.setOnMenuItemClickListener { item ->
      when (item.itemId) {
        R.id.move_top -> {
          MoveSong(entry, 0).defaultAction(holder.context)
          true
        }
        R.id.search_related -> {
          val intent = Intent(holder.context, SearchActivity::class.java)
              .putExtra(SearchActivity.INITIAL_QUERY_EXTRA, holder.entry?.song?.description)
              .putExtra(SearchActivity.INITIAL_PROVIDER_EXTRA, holder.entry?.song?.provider?.id)
          holder.context.startActivity(intent)
          true
        }
        R.id.remove_button -> {
          Thread({
            try {
              val token = Auth.apiKey.raw
              val newQueue = Connection.dequeue(
                  token, song.id, song.provider.id
              )
              QueueState.queue = newQueue
            } catch (e: ApiException) {
              if (e.code == 403) {
                Auth.clear()
              } else {
                e.printStackTrace()
                // TODO show toast
              }
            } catch (e: AuthException) {
              e.printStackTrace()
              // TODO show toast
            }
          }).start()
          true
        }
        else -> false
      }
    }
    holder.contextMenu.setOnClickListener {
      val apiKey = Auth.apiKeyNoRefresh
      val items = menu.menu
      items.findItem(R.id.move_top).isVisible = Auth.hasPermissionNoRefresh(Permission.MOVE)
      items.findItem(R.id.search_related).isVisible
      items.findItem(R.id.remove_button).isVisible =
          apiKey != null && apiKey.userName == entry.userName
              || Auth.hasPermissionNoRefresh(Permission.SKIP)
      menu.show()
    }
  }

  private fun TextView.setContent(content: String?) {
    if (content.isNullOrBlank()) {
      visibility = View.GONE
    } else {
      visibility = View.VISIBLE
      text = content
    }
  }

  override fun getItemCount(): Int = items.size

  internal fun getItem(position: Int): QueueEntry = items[position]

  class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
    val albumArtView: ImageView = view.album_art
    val titleView: TextView = view.song_title
    val descriptionView: TextView = view.song_description
    val durationView: TextView = view.song_duration
    val queuerView: TextView = view.song_queuer
    val contextMenu: ImageButton = view.context_menu
    var entry: QueueEntry? = null
    val context: Context
      get() = view.context

    init {
      this.descriptionView.isSelected = true
      this.titleView.isSelected = true
    }

    override fun toString(): String = "${super.toString()} '${titleView.text}'"
  }

  interface QueueEntryListener {
    fun onClick(entry: QueueEntry)
  }
}
